src/app_engine/gcmrecord.py

```diff
@@ -1,59 +1,159 @@
-#!/usr/bin/python2.4
+#!/usr/bin/python2.7
 #
 # Copyright 2015 Google Inc. All Rights Reserved.
 
-"""AppRTC gcm records
+"""Database model for AppRTC registration.
 
-This module implements the model for gcm records.
+GCMEncryptedRecord is the stored ndb model.
+GCMRecord is the decrypted version of the stored model which contains sensitive
+information that we can't store in plaintext. These exist in memory only.
+
+Some notes:
+- We need userId and gcmId as keys to look up associated records. These are
+  hashed so that they are not directly readable from db.
+
+- Database is vulnerable to rainbow attack if global salt is known. This will
+  expose user ids since phone numbers are easily iterable. Also, clearly db
+  is vulnerable if encryption key is known.
+
+- Also note that hashing gcmId to be used as a lookup key is not good since now
+  the encryption we use is only as good as the salt. However, in order to change
+  this we'd need to change the call APIs to provide both userId and gcmId where
+  they now take gcmId. In that model we would make all queries off the hashed
+  userId, and decrypt gcmId as needed. Potential future work.
 """
 
-import collections
-import constants
-import datetime
+import json
 import logging
-import util
+import time
+
+import constants
+from encryptor import Encryptor
+from encryptor import Hasher
 
 from google.appengine.ext import ndb
 
+
 def get_ancestor_key():
-  return ndb.Key('GCMRecord', 'global')
+  return ndb.Key('GCMEncryptedRecord', 'global')
+
 
-class GCMRecord(ndb.Model):
-  """Models a user ID to GCM ID binding record."""
-  user_id = ndb.StringProperty()
-  gcm_id = ndb.StringProperty()
-  code = ndb.StringProperty()
-  code_sent_time = ndb.DateTimeProperty()
-  verified = ndb.BooleanProperty()
-  last_modified_time = ndb.DateTimeProperty()
+class GCMEncryptedRecord(ndb.Model):
+  """Stores an encrypted registration in db."""
+  user_key = ndb.BlobProperty(indexed=True)
+  gcm_key = ndb.BlobProperty(indexed=True)
+  content = ndb.BlobProperty()
```

Comment:
Please comment that content is the encrypted json representation of GCMRecord, user_key is the hash of user id, gcm_key is the hash of gcm id.